server.js

```diff
@@ -1,23 +1,77 @@
 const express = require("express");
 const cors = require("cors");
-
 const app = express();
+app.use(express.json()); // before our routes definition
 
 app.use(cors());
+const bodyParser = require("body-parser");
+app.use(bodyParser.urlencoded({ extended: true }));
 
-const welcomeMessage = {
-  id: 0,
-  from: "Bart",
-  text: "Welcome to CYF chat system!",
-};
-
+let count = 0;
 //This array is our "data store".
 //We will start with one message in the array.
 //Note: messages will be lost when Glitch restarts our server.
-const messages = [welcomeMessage];
+const messages = [{ id: 20, from: "Alaa", text: "sd" }];
 
 app.get("/", function (request, response) {
   response.sendFile(__dirname + "/index.html");
 });
+// show all massages
+app.get("/messages", (res, resp) => {
+  resp.json({ messages });
+});
+app.get("/messages/search", (req, res) => {
+  const sText = req.query.text;
+  const matched = messages.filter((message) => {
+    const resultSearch = message.text.includes(sText);
+    return resultSearch;
+  });
+  console.log(matched);
+  res.json({ matched });
+});
+app.get("/messages/:id", (req, resp) => {
+  const massageId = Number(req.params.id);
+  const showMassage = messages.filter((massge) => massge.id === massageId);
+  resp.json(showMassage);
+});
 
-app.listen(process.env.PORT);
+app.delete("/messages/:id", (req, res) => {
+  const massageId = Number(req.params.id);
+  messages = messages.filter((mass) => mass.id !== massageId);
+});
+
+// app.get("/messages/search", (req, rep) => {
+//   const sText = req.query.search;
+//   console.log("keyword:", sText);
+//   const matched = messages.filter((message) => {
+//     message.text.toLowerCase().includes(sText.toLocaleLowerCase());
+//   });
+//   rep.send(matched);
+// });
+
+// add new massages
+app.post("/messages", (rec, resp) => {
+  const newMasseg = rec.body;
+  // input valdation
+  if (
+    newMasseg.from === "" ||
+    newMasseg.from === undefined ||
+    newMasseg.text === "" ||
+    newMasseg.text === undefined
+  ) {
```

Comment:
This if statement could be much cleaner if you used variables:
```
const fromValidation = newMessage.from === "" || newMessage.from === undefined
const textValidation = newMessage.text === "" || newMessage.text === undefined

if(fromValidation || textValidation){
     resp.send("Missing Information >>");
}
```

For clarity and readability I would also probably switch the cases around like so: 
```
const fromValidation = newMessage.from !== "" || newMessage.from !== undefined
const textValidation = newMessage.text !== "" || newMessage.text !== undefined

if(fromValidation || textValidation){
     let userMessage = {
      id: count++,
      from: newMessage.from,
      text: newMessage.text,
    };
    messages.push(userMessage);
    resp.json({ userMessage });
} else {
     resp.send("Missing Information >>");
}
```
Can you see how I am looking for NOT `""` and NOT `undefined` and then doing the `newMessage` logic in here. Its generally easier to read a NOT -> flow than vice versa. LMK any questions.